File: app/src/main/java/com/android/popularmoviesapp/MovieDetailActivity.java
package com.android.popularmoviesapp;

import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.database.DatabaseUtils;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.LoaderManager;
import android.support.v4.app.NavUtils;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.databinding.DataBindingUtil;
import android.view.View;
import android.widget.ImageButton;

import com.android.popularmoviesapp.data.MovieData;
import com.android.popularmoviesapp.data.MovieDbHelper;
import com.android.popularmoviesapp.databinding.ActivityMovieDetailBinding;

import com.android.popularmoviesapp.data.MovieContract;
import com.android.popularmoviesapp.sync.FavoritesMovieIntentService;
import com.android.popularmoviesapp.sync.MovieInfo;
import com.android.popularmoviesapp.sync.MovieReviewIntentService;
import com.android.popularmoviesapp.utilities.MovieDatabaseJsonUtils;
import com.android.popularmoviesapp.utilities.NetworkUtils;
import com.squareup.picasso.Picasso;

import java.net.URL;
import java.util.ArrayList;

import static com.android.popularmoviesapp.utilities.NetworkUtils.getResponseFromHttpUrl;

public class MovieDetailActivity extends AppCompatActivity implements
        LoaderManager.LoaderCallbacks<ArrayList<MovieData>>{

    final String TAG = MovieDetailActivity.class.getSimpleName();

    public static final String[] MOVIE_DETAIL_PROJECTION = {
            MovieContract.MovieEntry.COLUMN_TITLE,
            MovieContract.MovieEntry.COLUMN_VOTE_AVERAGE,
            MovieContract.MovieEntry.COLUMN_POSTER_PATH,
            MovieContract.MovieEntry.COLUMN_BACKDROP_PATH,
            MovieContract.MovieEntry.COLUMN_RELEASE_DATE,
            MovieContract.MovieEntry.COLUMN_OVERVIEW,
            MovieContract.MovieEntry.COLUMN_MOVIE_ID,
            MovieContract.MovieEntry.COLUMN_TRAILER_KEY,
            MovieContract.MovieEntry.COLUMN_TRAILER_TYPE,
            MovieContract.MovieEntry.COLUMN_TRAILER_NAME,
            MovieContract.MovieEntry.COLUMN_REVIEW_AUTHOR,
            MovieContract.MovieEntry.COLUMN_REVIEW_CONTENT,
            MovieContract.MovieEntry.COLUMN_REVIEW_URL,
            MovieContract.MovieEntry.COLUMN_FAVORITE_BOOL
    };

    public static final int INDEX_MOVIE_TITLE = 0;
    public static final int INDEX_MOVIE_VOTE_AVERAGE = 1;
    public static final int INDEX_MOVIE_POSTER_PATH = 2;
    public static final int INDEX_MOVIE_BACKDROP_PATH = 3;
    public static final int INDEX_MOVIE_RELEASE_DATE = 4;
    public static final int INDEX_MOVIE_OVERVIEW = 5;
    public static final int INDEX_MOVIE_ID = 6;

    public static final int INDEX_TRAILER_KEY = 7;
    public static final int INDEX_TRAILER_TYPE = 8;
    public static final int INDEX_TRAILER_NAME = 9;

    public static final int INDEX_REVIEW_AUTHOR = 10;
    public static final int INDEX_REVIEW_CONTENT = 11;
    public static final int INDEX_REVIEW_URL = 12;
    public static final int INDEX_FAVORITE_BOOL = 13;

    private static final int ID_MOVIE_TRAILER_LOADER = 519;
    private static final int ID_MOVIE_REVIEW_LOADER = 119;


    //private Uri mUri;
    //private Cursor mCursor;
    private Context mContext = MovieDetailActivity.this;

    private ActivityMovieDetailBinding mDetailBinding;

    private RecyclerView mMovieTrailerList;
    private RecyclerView mMovieReviewList;

    private MovieTrailerAdapter mTrailerAdapter;
    private MovieReviewAdapter mReviewAdapter;

    private int mPosition = RecyclerView.NO_POSITION;

    private boolean checkFavorite = false;

    FloatingActionButton favoriteFab;
    SQLiteDatabase mDb;

    MovieData movie_details;

    @Override
    protected void onStart() {
        super.onStart();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_movie_detail);

        mDetailBinding = DataBindingUtil.setContentView(this,
                R.layout.activity_movie_detail);

        favoriteFab = mDetailBinding.favoriteFloatingActionButton;

        ActionBar actionBar = this.getSupportActionBar();
        if(actionBar != null){
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        RecyclerView.LayoutManager layoutTrailerManager =
                new LinearLayoutManager(getApplicationContext());

        RecyclerView.LayoutManager layoutReviewManager =

                new LinearLayoutManager(getApplicationContext());

        mMovieTrailerList = mDetailBinding.mdTrailersRv;
        mMovieReviewList = mDetailBinding.mdReviewsRv;

        mMovieTrailerList.setLayoutManager(layoutTrailerManager);
        mMovieReviewList.setLayoutManager(layoutReviewManager);

        Intent intent = getIntent();
        if (intent == null) throw new NullPointerException("YOUR INTENT cannot be null");
        movie_details = (MovieData) intent.getSerializableExtra("movieDeets");

        String movieTitle = movie_details.getOriginal_title();

        String moviePoster = movie_details.getPoster_path();

        String movieBackdrop = movie_details.getBackdrop_path();

        String movieOverview = movie_details.getOverview();

        String movieReleaseDate = movie_details.getRelease_date();

        String movieRating = movie_details.getVote_average();

        //create the image url for poster and backdrop
        final String IMAGE_BASE = "http://image.tmdb.org/t/p/";
        final String IMAGE_SIZE_W500 = "w500";
        final String IMAGE_SIZE_ORIGINAL = "original";

        //obtain poster path from MovieData object
        String POSTER_IMAGE_URL = moviePoster;
        String BACKDROP_IMAGE_URL = movieBackdrop;

        String poster_url = IMAGE_BASE + IMAGE_SIZE_W500 + POSTER_IMAGE_URL;
        String backdrop_url = IMAGE_BASE + IMAGE_SIZE_ORIGINAL + BACKDROP_IMAGE_URL;
        //output the url into Log for debug purposes
        //Log.d(TAG, trailer_KEY);

        //use Picasso to place url image into imageview
        Picasso.get()
                .load(poster_url)
                .into(mDetailBinding.movieDetailIv);
        Picasso.get()
                .load(backdrop_url)
                .into(mDetailBinding.movieDetailBackdropIv);

        if (checkFavorite){
            checkFavorite = !checkFavorite;
            favoriteFab.setImageDrawable(ContextCompat.getDrawable(getApplicationContext(), R.drawable.ic__favorite_movie_action_on));
        } else {
            checkFavorite = !checkFavorite;
            favoriteFab.setImageDrawable(ContextCompat.getDrawable(getApplicationContext(), R.drawable.ic__favorite_movie_action_off));
        }

        favoriteFab.setOnClickListener(new View.OnClickListener() {
           @Override
           public void onClick(View view) {
               if (checkFavorite) {
                   checkFavorite = !checkFavorite;
                   //TODO: find a way to write boolean value to cursor
                   favoriteFab.setImageDrawable(ContextCompat.getDrawable(getApplicationContext(),
                           R.drawable.ic__favorite_movie_action_off));
               } else {
                   checkFavorite = !checkFavorite;
                   favoriteFab.setImageDrawable(ContextCompat.getDrawable(getApplicationContext(),
                           R.drawable.ic__favorite_movie_action_on));

               }
           }
        });

        mDetailBinding.mdOriginalTitle.setText(movieTitle);
        mDetailBinding.mdVoteAvg.setText(movieRating);
        mDetailBinding.mdReleaseDate.setText(movieReleaseDate);
        mDetailBinding.mdPlotSynopsis.setText(movieOverview);

        //mDetailBinding.mdOriginalTitleLabel.setText(getString(R.string.md_original_title_label));
        mDetailBinding.mdPlotSynopsisLabel.setText(getString(R.string.md_synopsis_label));
        mDetailBinding.mdReleaseDateLabel.setText(getString(R.string.md_date_label));
        mDetailBinding.mdVoteAvgLabel.setText(getString(R.string.md_rating_label));

        String movie_ID = movie_details.getMovie_id();
        System.out.println(movie_ID);

        getSupportLoaderManager().initLoader(ID_MOVIE_TRAILER_LOADER, null, this);

    }

    //will set BACK BUTTON to navigate back to previous activity w/o recreating it.
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemClicked = item.getItemId();
        Context context = MovieDetailActivity.this;


        switch (itemClicked) {

            case android.R.id.home:
                NavUtils.navigateUpFromSameTask(this);

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @NonNull
    @Override
    public Loader<ArrayList<MovieData>> onCreateLoader(int loaderId, @Nullable Bundle args) {

        switch (loaderId) {

            case ID_MOVIE_TRAILER_LOADER:
                System.out.println(loaderId);

                return new AsyncTaskLoader<ArrayList<MovieData>>(this) {
                    @Override
                    protected void onStartLoading() {
                        super.onStartLoading();
                        forceLoad();
                    }

                    @Nullable
                    @Override
                    public ArrayList<MovieData> loadInBackground() {
                        String movie_ID = movie_details.getMovie_id();
                        System.out.println("ASYNC LOAD STARTED: " + movie_ID);
                        try{
                            URL trailerRequestUrl = NetworkUtils.trailersBuildUrl(movie_ID);

                            String jsonMovieDatabaseResponse = getResponseFromHttpUrl(trailerRequestUrl);

                            ArrayList<MovieData> simpleJsonTrailerData = MovieDatabaseJsonUtils.getMovieTrailerData(jsonMovieDatabaseResponse);

                            Log.v(TAG, "ASYNCTASK: " + simpleJsonTrailerData.get(0).getTrailer_type());

                            return simpleJsonTrailerData;

                        } catch (Exception e) {
                            e.printStackTrace();
                            return null;
                        }
                    }
                };

            case ID_MOVIE_REVIEW_LOADER:

                return new AsyncTaskLoader<ArrayList<MovieData>>(this) {
                    @Override
                    protected void onStartLoading() {
                        super.onStartLoading();
                        forceLoad();
                    }

                    @Nullable
                    @Override
                    public ArrayList<MovieData> loadInBackground() {
                        String movie_ID = movie_details.getMovie_id();
                        System.out.println("ASYNC LOAD STARTED: " + movie_ID);
                        try {
                            URL movieRequestUrl = NetworkUtils.reviewsBuildUrl(movie_ID);
                            String jsonMDBResponse = getResponseFromHttpUrl(movieRequestUrl);

                            ArrayList<MovieData> simpleJsonReviewData = MovieDatabaseJsonUtils.getMovieReviewData(jsonMDBResponse);

                            Log.v(TAG, "ASYNCTASK: " + simpleJsonReviewData.get(0).getReview_author());

                            return simpleJsonReviewData;

                        } catch (Exception e) {
                            e.printStackTrace();
                            return null;
                        }
                    }
                };
            default:
                throw new RuntimeException("Loader Not Implemented: "  + loaderId);
        }
    }

    @Override
    public void onLoadFinished(@NonNull Loader<ArrayList<MovieData>> loader, ArrayList<MovieData> data) {

        if(loader.getId() == ID_MOVIE_TRAILER_LOADER) {
            getSupportLoaderManager().restartLoader(ID_MOVIE_REVIEW_LOADER, null, this);
            //System.out.println("TRAILER load finished " + data.get(0).getTrailer_key());
            mTrailerAdapter = new MovieTrailerAdapter(5,
                    getApplicationContext(), data);
            mMovieTrailerList.setAdapter(mTrailerAdapter);
        }


        if (loader.getId() == ID_MOVIE_REVIEW_LOADER){
            //System.out.println("REVIEW load finished " + data.get(0).getReview_author());

            mReviewAdapter = new MovieReviewAdapter(getApplicationContext(), data);
            mMovieReviewList.setAdapter(mReviewAdapter);
        }

    }

    @Override
    public void onLoaderReset(@NonNull Loader<ArrayList<MovieData>> loader) {
        //mTrailerAdapter.swapCursor(null);
        //mReviewAdapter.swapCursor(null);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.movie_detail_menu, menu);
        return true;
    }

}

File: app/src/main/java/com/android/popularmoviesapp/MovieDetailActivity.kt
package com.android.popularmoviesapp

import android.content.ContentValues
import android.database.sqlite.SQLiteDatabase
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.LoaderManager
import android.support.v4.app.NavUtils
import android.support.v4.content.AsyncTaskLoader
import android.support.v4.content.ContextCompat
import android.support.v4.content.Loader
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.MenuItem
import com.android.popularmoviesapp.data.MovieContract
import com.android.popularmoviesapp.data.MovieData
import com.android.popularmoviesapp.data.MovieDbHelper
import com.android.popularmoviesapp.databinding.ActivityMovieDetailBinding
import com.android.popularmoviesapp.utilities.MovieDatabaseJsonUtils
import com.android.popularmoviesapp.utilities.NetworkUtils
import com.squareup.picasso.Picasso
import java.util.*


class MovieDetailActivity : AppCompatActivity(), LoaderManager.LoaderCallbacks<ArrayList<MovieData>> {

    internal val TAG = MovieDetailActivity::class.java.simpleName

    private val mContext = this@MovieDetailActivity

    private var mDetailBinding: ActivityMovieDetailBinding? = null

    private var mMovieTrailerList: RecyclerView? = null
    private var mMovieReviewList: RecyclerView? = null

    private var mTrailerAdapter: MovieTrailerAdapter? = null
    private var mReviewAdapter: MovieReviewAdapter? = null

    private var checkFavorite: Boolean = false
    private lateinit var favoriteFab: FloatingActionButton
    private lateinit var mDb: SQLiteDatabase

    private lateinit var movieDetails: MovieData

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_movie_detail)

        mDetailBinding = DataBindingUtil.setContentView(this,
                R.layout.activity_movie_detail)

        favoriteFab = mDetailBinding!!.favoriteFloatingActionButton

        val movieDbHelper = MovieDbHelper(this)
        mDb = movieDbHelper.writableDatabase

        val actionBar = this.supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)

        val layoutTrailerManager = LinearLayoutManager(applicationContext)

        val layoutReviewManager =

                LinearLayoutManager(applicationContext)

        mMovieTrailerList = mDetailBinding!!.mdTrailersRv
        mMovieReviewList = mDetailBinding!!.mdReviewsRv

        mMovieTrailerList!!.layoutManager = layoutTrailerManager
        mMovieReviewList!!.layoutManager = layoutReviewManager

        val intent = intent ?: throw NullPointerException("YOUR INTENT cannot be null")
        movieDetails = intent.getSerializableExtra("movieDeets") as MovieData

        val movieTitle = movieDetails.original_title

        val moviePoster = movieDetails.poster_path

        val movieBackdrop = movieDetails.backdrop_path

        val movieOverview = movieDetails.overview

        val movieReleaseDate = movieDetails.release_date

        val movieRating = movieDetails.vote_average

        //create the image url for poster and backdrop
        val  imageBaseURL = "http://image.tmdb.org/t/p/"
        val imageSizeW500 = "w500"
        val imageSizeOriginal = "original"

        //obtain poster path from MovieData object

        val poster_url = imageBaseURL + imageSizeW500 + moviePoster
        val backdrop_url = imageBaseURL + imageSizeOriginal + movieBackdrop

        //use Picasso to place url image into imageview
        Picasso.get()
                .load(poster_url)
                .placeholder(R.drawable.ic_blank_black)
                .error(R.drawable.ic_sentiment_very_dissatisfied_black_24dp)
                .into(mDetailBinding!!.movieDetailIv)
        Picasso.get()
                .load(backdrop_url)
                .placeholder(R.drawable.ic_blank_black)
                .error(R.drawable.ic_sentiment_very_dissatisfied_black_24dp)
                .into(mDetailBinding!!.movieDetailBackdropIv)

        if (movieDetails.check_favorite_movie != null) {
            checkFavorite = java.lang.Boolean.parseBoolean(movieDetails.check_favorite_movie)
        } else {
            checkFavorite = false
        }

        if (checkFavorite) {
            favoriteFab.setImageDrawable(ContextCompat.getDrawable(applicationContext, R.drawable.ic__favorite_movie_action_on))

        } else {
            favoriteFab.setImageDrawable(ContextCompat.getDrawable(applicationContext, R.drawable.ic__favorite_movie_action_off))
        }

        favoriteFab.setOnClickListener {
            if (checkFavorite) {
                favoriteFab.setImageDrawable(ContextCompat.getDrawable(applicationContext,
                        R.drawable.ic__favorite_movie_action_off))
                //delete happens here
                mDb.delete(MovieContract.MovieEntry.TABLE_NAME_MOVIE_MAIN, "id=?", arrayOf(movieDetails.movie_id))
                checkFavorite = !checkFavorite

            } else {
                favoriteFab.setImageDrawable(ContextCompat.getDrawable(applicationContext,
                        R.drawable.ic__favorite_movie_action_on))
                checkFavorite = !checkFavorite
                addFavoriteMovie()
            }
        }

        mDetailBinding!!.mdOriginalTitle.text = movieTitle
        mDetailBinding!!.mdVoteAvg.text = movieRating
        mDetailBinding!!.mdReleaseDate.text = movieReleaseDate
        mDetailBinding!!.mdPlotSynopsis.text = movieOverview

        //mDetailBinding.mdOriginalTitleLabel.setText(getString(R.string.md_original_title_label));
        mDetailBinding!!.mdPlotSynopsisLabel.text = getString(R.string.md_synopsis_label)
        mDetailBinding!!.mdReleaseDateLabel.text = getString(R.string.md_date_label)
        mDetailBinding!!.mdVoteAvgLabel.text = getString(R.string.md_rating_label)

        val movieId = movieDetails.movie_id
        println(movieId)

        supportLoaderManager.initLoader(ID_MOVIE_TRAILER_LOADER, null, this)
    }

    //will set BACK BUTTON to navigate back to previous activity w/o recreating it.
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemClicked = item.itemId

        when (itemClicked) {

            android.R.id.home -> {
                NavUtils.navigateUpFromSameTask(this)
                return super.onOptionsItemSelected(item)
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateLoader(loaderId: Int, args: Bundle?): Loader<ArrayList<MovieData>> {

        when (loaderId) {

            ID_MOVIE_TRAILER_LOADER -> {
                println(loaderId)

                return object : AsyncTaskLoader<ArrayList<MovieData>>(this) {
                    var mMovieData: ArrayList<MovieData>? = null

                    override fun onStartLoading() {
                        if (mMovieData != null) {
                            deliverResult(mMovieData)
                            println("TRAILER DATA! delivering data!")
                        } else {
                            forceLoad()
                            println("TRAILER DATA forceloading.")
                        }
                    }

                    override fun deliverResult(data: ArrayList<MovieData>?) {
                        mMovieData = data
                        super.deliverResult(data)
                    }

                    override fun loadInBackground(): ArrayList<MovieData>? {
                        val movieId = movieDetails.movie_id
                        println("ASYNC TRAILER LOAD STARTED: " + movieId!!)
                        try {
                            val trailerRequestUrl = NetworkUtils.trailersBuildUrl(movieId)

                            val jsonMovieDatabaseResponse = NetworkUtils.getResponseFromHttpUrl(trailerRequestUrl!!)

//Log.v(TAG, "ASYNCTASK: " + simpleJsonTrailerData.get(0).getTrailer_type());

                            return MovieDatabaseJsonUtils.getMovieTrailerData(jsonMovieDatabaseResponse!!)

                        } catch (e: Exception) {
                            e.printStackTrace()
                            return null
                        }

                    }
                }
            }

            ID_MOVIE_REVIEW_LOADER ->

                return object : AsyncTaskLoader<ArrayList<MovieData>>(this) {
                    var mMovieData: ArrayList<MovieData>? = null

                    override fun onStartLoading() {
                        if (mMovieData != null) {
                            deliverResult(mMovieData)
                            println("REVIEW DATA DELIVERED")
                        } else {
                            forceLoad()
                            println("REVIEW DATA FORCELOADED.")
                        }
                    }

                    override fun deliverResult(data: ArrayList<MovieData>?) {
                        mMovieData = data
                        super.deliverResult(data)
                    }

                    override fun loadInBackground(): ArrayList<MovieData>? {
                        val movieId = movieDetails.movie_id
                        println("ASYNC REVIEW LOAD STARTED: " + movieId!!)
                        try {
                            val movieRequestUrl = NetworkUtils.reviewsBuildUrl(movieId)
                            val jsonMDBResponse = NetworkUtils.getResponseFromHttpUrl(movieRequestUrl!!)

//Log.v(TAG, "ASYNCTASK: " + simpleJsonReviewData.get(0).getReview_author());

                            return MovieDatabaseJsonUtils.getMovieReviewData(jsonMDBResponse!!)

                        } catch (e: Exception) {
                            e.printStackTrace()
                            return null
                        }

                    }
                }

            else -> throw RuntimeException("Loader Not Implemented: $loaderId")
        }
    }

    override fun onLoadFinished(loader: Loader<ArrayList<MovieData>>, data: ArrayList<MovieData>) {

        if (loader.id == ID_MOVIE_TRAILER_LOADER) {
            //System.out.println("TRAILER load finished " + data.get(0).getTrailer_key());
            mTrailerAdapter = MovieTrailerAdapter(applicationContext, data)
            mMovieTrailerList!!.adapter = mTrailerAdapter
            supportLoaderManager.initLoader(ID_MOVIE_REVIEW_LOADER, null, this)
        }

        if (loader.id == ID_MOVIE_REVIEW_LOADER) {
            //System.out.println("REVIEW load finished " + data.get(0).getReview_author());

            mReviewAdapter = MovieReviewAdapter(applicationContext, data)
            mMovieReviewList!!.adapter = mReviewAdapter
        }

    }

    override fun onLoaderReset(loader: Loader<ArrayList<MovieData>>) {

    }

    private fun addFavoriteMovie() {
        val updateFavoriteMovie = ContentValues()
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_MOVIE_ID, movieDetails.movie_id)
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_TITLE, movieDetails.original_title)
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_POSTER_PATH, movieDetails.poster_path)
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_BACKDROP_PATH, movieDetails.backdrop_path)
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_OVERVIEW, movieDetails.overview)
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_RELEASE_DATE, movieDetails.release_date)
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_VOTE_AVERAGE, movieDetails.vote_average)
        updateFavoriteMovie.put(MovieContract.MovieEntry.COLUMN_FAVORITE_BOOL, java.lang.Boolean.toString(checkFavorite))
        Log.v(TAG, movieDetails.original_title + " " + checkFavorite)

        val insertCount = mDb.insert(MovieContract.MovieEntry.TABLE_NAME_MOVIE_MAIN, null, updateFavoriteMovie)
    }

    companion object {

        private const val ID_MOVIE_TRAILER_LOADER = 519
        private const val ID_MOVIE_REVIEW_LOADER = 119
    }

}
